BSP/Src/BSP_I2C.c

```diff
@@ -0,0 +1,214 @@
+ */
+void BSP_I2C_RegisterWrite(I2C_HandleTypeDef *hi2c, 
+              uint32_t deviceAddress, 
+              uint32_t memoryAddress, //register
+              uint32_t memoryAddressSize,
+              uint8_t* pDataBuff, 
+              uint16_t len) {
+    //copy buffer data over for security incase it changes 
+    uint8_t buffer[I2C_QUEUE_SIZE * I2C_ITEM_LENGTH];
+        for (uint16_t i = 0; i < len; i++) {
+            buffer[i] = pDataBuff[i];
+        }
+    xSemaphoreTake(semaphore, 1);
+    if (!__HAL_I2C_GET_FLAG(hi2c, I2C_FLAG_BUSY)) {
+        // HAL requires 7b I2C address left shifted
+        HAL_I2C_Mem_Write_IT(hi2c, deviceAddress << 1, memoryAddress, memoryAddressSize, buffer, len);
+    }
+    else if (uxQueueSpacesAvailable(I2C_DataRegisterQueue) >= len) {
+        for (uint32_t i = 0; i < len; i++) {
+            xQueueSend(I2C_DataRegisterQueue, buffer[i], 0);
+        }
+            // HAL requires 7b I2C address left shifted
+            xQueueSend(I2C_MetaRegisterQueue, deviceAddress << 1, 0);
+            xQueueSend(I2C_MetaRegisterQueue, memoryAddress, 0);
+            xQueueSend(I2C_MetaRegisterQueue, memoryAddressSize, 0);
+            xQueueSend(I2C_MetaRegisterQueue, len, 0);
+    }
+    xSemaphoreGive(semaphore);
+}
+
+/**
+ * @brief Recursively go through to see if there is any more info that needs to be transmitted,
+ * if there is transmit it
+ * @param hi2c                  I2C handle that interrupted (passed in by IRQ)
+ */
+void HAL_I2C_MemTxCpltCallback(I2C_HandleTypeDef *hi2c) {
+    uint8_t buffer[I2C_QUEUE_SIZE * I2C_ITEM_LENGTH];
+    if (xQueueIsQueueEmptyFromISR(I2C_MetaRegisterQueue) == pdFALSE) {
+        uint32_t deviceAddress, memoryAddress, memoryAddressSize, len;
+        uint8_t temp;
+        xQueueReceiveFromISR(I2C_MetaRegisterQueue, &deviceAddress, &temp);
+        xQueueReceiveFromISR(I2C_MetaRegisterQueue, &memoryAddress, &temp);
+        xQueueReceiveFromISR(I2C_MetaRegisterQueue, &memoryAddressSize, &temp);
+        xQueueReceiveFromISR(I2C_MetaRegisterQueue, &len, &temp);
+        for (uint32_t i = 0; i < len; i++) {
+            xQueueReceiveFromISR(I2C_DataRegisterQueue, &(buffer[i]), &temp);
+        }
+        HAL_I2C_Mem_Write_IT(hi2c, deviceAddress, memoryAddress, memoryAddressSize, buffer, len);
+    }
+}
+
+/**
+ * @brief    Gets the data from a device through the I2C bus.
+ * @param    hi2c               struct with config data
+ * @param    deviceAddress      target device address
+ * @param    memoryAddress      the register address to write to in the IC's memory.
+ * @param    memoryAddressSize  the register address size
+ * @return   void
+ */
+//Register Read
+void BSP_I2C_Read(I2C_HandleTypeDef* hi2c,
+              uint16_t deviceAddress,
+              uint16_t memoryAddress,
+              uint16_t memoryAddressSize) {
+    HAL_I2C_Mem_Read_IT(hi2c, deviceAddress, memoryAddress, sizeof(uint8_t), __ReceiveRawData, I2C_QUEUE_SIZE * I2C_ITEM_LENGTH);
```

Comment:
return status here too.

Reply:
problem with the async read initialized from the hal mem read it callback is that messages will be missed because the caller needs to call this function prior to receiving; also the caller won't know when this finishes